src/mbgl/map/map_impl.cpp

```diff
@@ -172,4 +172,14 @@ void Map::Impl::jumpTo(const CameraOptions& camera) {
     onUpdate();
 }
 
+void Map::Impl::onStyleImageMissing(const std::string& id, RendererObserver::StyleImageMissingCallback done) {
+
+    if (style->getImage(id) == nullptr) {
+        observer.onStyleImageMissing(id);
+    }
+
+    done();
```

Comment:
Why is there a `done` function? What does it do that we can't do just through the mere fact of returning from this function?

Reply:
I might be wrong here, but: on Android the map and renderer are on different threads. The missing image event originates on the render thread but needs to be called on the ui thread. [The observer for `RendererFrontend`](https://github.com/mapbox/mapbox-gl-native/blob/2fed2155c58bf8dfaa00062928f766fd848163a0/include/mbgl/renderer/renderer_frontend.hpp#L23-L25) needs to be implemented in a way to calls onto the main thread and then calls the callback back on the render thread. I haven't implemented the changes to the Android platform yet.

Reply:
Wait, I think you're right. This isn't necessary here

Reply:
So, I think the `RendererFrontend`'s observer needs to have this callback and since `Map::Impl` implements the same interface it also needs it